pkg/storegateway/sharding_strategy.go

```diff
@@ -170,6 +171,18 @@ func (s *ShuffleShardingStrategy) FilterBlocks(_ context.Context, userID string,
 // GetShuffleShardingSubring returns the subring to be used for a given user. This function
 // should be used both by store-gateway and querier in order to guarantee the same logic is used.
 func GetShuffleShardingSubring(ring *ring.Ring, userID string, limits ShardingLimits) ring.ReadRing {
+	// Check per-zone shard size first. When set, the total shard size is computed as
+	// per-zone shard size * number of zones. This is useful during zone migrations
+	// (e.g., RF 3 -> RF 4) because it ensures the shard size scales with the number
+	// of zones, preventing block reshuffling in existing zones.
+	if shardSizePerZone := limits.StoreGatewayTenantShardSizePerZone(userID); shardSizePerZone > 0 {
```

Comment:
[nit] I would refactor this entire block to simply compute the `shardSize` and then let the rest of the function continue. This way we call `ShuffleShard()` only once in the function, with the right `shardSize`.